### PR TITLE
Knawat Marketplace

### PR DESCRIPTION
**Pull request by Abdelrahman3D**
This *PR* mainly aims to refactor the code base, split it into different modules, add both basic and bearer auth support.

### CODE REVIEW

File: src/KnawatMP.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+      }),
+    }).then(({ channel }) => {
+      this.token = channel.token;
+      return channel.token;
+    });
+  }
+
+  async setAuthHeaders(authType) {
+    if (authType === 'none') {
+      this.headers.Authorization = '';
```
Comment (ElGreetly) [inline src/KnawatMP.js]: you may delete the Authorization here better?!

File: src/KnawatMP.js [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+      return;
+    }
+    if (authType === 'basic') {
+      this.headers.Authorization = `Basic ${this.getBasicAuth()}`;
+      return;
```
Comment (ElGreetly) [inline src/KnawatMP.js]: why did you add return?

File: src/api/orders.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
+import querystring from 'querystring';
```
Comment (ElGreetly) [inline src/api/orders.js]: I think this could be implemented inside $fetch?

File: src/KnawatMP.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    return fetch(`${KnawatMP.baseUrl}${path}`, {
+      method: method,
+      headers: this.headers,
+      ...options,
+    }).then(async res => {
+      const jsonRes = await res.json();
+      if (res.ok) {
+        return jsonRes;
+      }
+      throw jsonRes;
```
Comment (ElGreetly) [inline src/KnawatMP.js]: you will add the statusCode?

Comment (Abdelrahman3D) [reply src/KnawatMP.js]: Yes, I will.

File: src/KnawatMP.js (hunk 1 above)
Comment (Abdelrahman3D) [reply src/KnawatMP.js]: Yes for sure

File: src/api/orders.js (hunk 3 above)
Comment (Abdelrahman3D) [reply src/api/orders.js]: I will, and I will use `qs` package instead of `querystring`

File: src/api/categories.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+  /**
+   * get invoices list
+   * @param {object} args
+   */
+  getCategories(args = {}) {
+    const params = Object.entries(args)
+      .map(([key, val]) => `${key}=${val}`)
+      .join('&');
+    return this.$fetch('GET', `/catalog/categories?${params}`, { auth: 'token' });
```
Comment (ElGreetly) [inline src/api/categories.js]: you already implemented qs in the general $fetch why don't you send the queryParams on the options?

File: src/api/products.js [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+   * @param {object} { limit = 10, page = 1, lastUpdate = null, keyword = null, hideOutOfStock = 0 }
+   * @returns
+   * @see https://knawat-mp.restlet.io/#operation_get_products
+   * @memberof MP
+   */
+  getProducts({
+    limit = 10,
+    page = 1,
+    lastUpdate = null,
+    keyword = null,
```
Comment (ElGreetly) [inline src/api/products.js]: why the default here is null?

File: src/api/membership.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+      body: JSON.stringify(body),
+      auth: 'basic',
+    });
+  },
+
+  /**
+   *
+   * @param {*} body
+   */
+  getCoupon({ membership, coupon }) {
```
Comment (ElGreetly) [inline src/api/membership.js]: this will create an issue please fix it if membership is undefined
you also can put coupon first as it's required then membership 
(coupon, membership)

File: src/api/membership.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+  addSubscription(body) {
+    return this.$fetch('POST', '/subscription', {
+      body: JSON.stringify(body),
+      auth: 'basic',
+    });
+  },
+
+  /**
+   *
+   * @param {*} body
```
Comment (ElGreetly) [inline src/api/membership.js]: docs

File: src/api/membership.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+   * @returns
+   * @memberof MP
+   */
+  getMembershipById(id) {
+    return this.$fetch('GET', `/membership/${id}`, { auth: 'basic' });
+  },
+
+  /**
+   *
+   * @param {*} body
```
Comment (ElGreetly) [inline src/api/membership.js]: docs

File: src/api/invoices.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+export default {
+  getInvoices() {
```
Comment (ElGreetly) [inline src/api/invoices.js]: docs

File: src/KnawatMP.js [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+  /**
+   * Fetch data from server
+   *
+   * @param {string} method
+   * @param {string} path
+   * @param {object} options
+   * @memberof Fetch
+   */
+  async $fetch(method, path, options = {}) {
```
Comment (ElGreetly) [inline src/KnawatMP.js]: options could be specified like { queryParams, auth } or add expected attributes to the docs

File: src/KnawatMP.js [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  getBasicAuth() {
+    const user = process.env.KNAWAT_USER;
+    const pass = process.env.KNAWAT_PASS;
+    if (!user || !pass) {
+      throw new Error('[KNAWAT MP] can not find a user or password');
+    }
+    return `${Buffer.from(`${user}:${pass}`).toString('base64')}`;
+  }
+
+  async getTokenAuth() {
```
Comment (ElGreetly) [inline src/KnawatMP.js]: docs

File: src/KnawatMP.js [hunk 13]
```diff
@@ -0,0 +1,102 @@
+import fetch from 'node-fetch';
+import qs from 'qs';
+
+class KnawatMP {
+  static baseUrl = process.env.MP_BASEURL || 'https://mp.knawat.io/api';
+  headers = {
+    'Content-Type': 'application/json',
+  };
+
+  static _addToPrototype(fns) {
```
Comment (ElGreetly) [inline src/KnawatMP.js]: docs

File: src/KnawatMP.js [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+        consumerKey: this.consumerKey,
+        consumerSecret: this.consumerSecret,
+      }),
+    }).then(({ channel }) => {
+      this.token = channel.token;
+      return channel.token;
+    });
+  }
+
+  async setAuthHeaders(authType) {
```
Comment (ElGreetly) [inline src/KnawatMP.js]: docs

File: src/api/shipping.js [hunk 15]
```diff
@@ -0,0 +1,25 @@
+export default {
+  /**
+   * get shipping fees based on weight and country
+   * @param {string} country
+   * @param {number} weight
+   */
+  getShippingRules(country, weight, price) {
+    const params = `?country=${country}&weight=${weight}&price=${price}`;
```
Comment (ElGreetly) [inline src/api/shipping.js]: send it as queryParams

File: src/api/stores.js [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+   */
+  getCurrentStore() {
+    return this.$fetch('GET', '/stores/me', { auth: 'basic' });
+  },
+
+  /**
+   * get shipping couriers
+   */
+  getStoreLogs(params) {
+    const searchParams = Object.entries(params).reduce((acc, [key, val]) => {
```
Comment (ElGreetly) [inline src/api/stores.js]: use qs => queryParams

File: src/api/categories.js [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+   */
+  getAlCategories() {
+    return this.$fetch('GET', '/catalog/categories', { auth: 'token' });
+  },
+
+  /**
+   * get invoices list
+   * @param {object} args
+   */
+  getCategories(args = {}) {
```
Comment (ElGreetly) [inline src/api/categories.js]: args is not clear what's the options?

File: src/api/stores.js [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+
+  /**
+   * get the current store data
+   */
+  getCurrentStore() {
+    return this.$fetch('GET', '/stores/me', { auth: 'token' });
+  },
+
+  /**
+   * re Sync store
```
Comment (ElGreetly) [inline src/api/stores.js]: just sync or resync
